Add CLI flags -n, -t, T and -p

AntoniosBarotsis opened the pull request:
This closes #13.

Only added @jhodapp as a reviewer because that's who I was in contact with when working on this, feel free to add anyone else!

The relevant slack thread can be found [here](https://rustneversleepshq.slack.com/archives/C02THGDG60Y/p1676392959503609).

## Added

```
  -n, --post-amount <POST_AMOUNT>
          The number of sensor readings to post. [DEFAULT: 1]
  -t, --time-per-post <TIME_PER_POST_S>
          The time between each sensor reading post (in seconds). [DEFAULT: 10]
  -T, --total-time <TOTAL_TIME_S>
          The total time over which all the sensor reading posts must be sent (in seconds, alternative to -t)
  -p, --num-threads <NUM_THREADS>
          The number of threads to spawn. The workload will be cloned to each thread, not divided. [DEFAULT: 1]
```

## Testing

From some *manual* testing, everything seems fine from what I can tell. It would be great if you could play around with it for a little bit to verify both that it works and that it *feels* as you'd expect (things like flag usage, default behavior, etc).

You can run it with `cargo r -- -h` or you can install it locally with `cargo install --path .` and then `ambi_mock_client -h`.

## Other Changes

This PR also upgraded to the latest version of clap and a bunch of other dependencies, moved a bunch of stuff around and added a logger.

AntoniosBarotsis commented:
This includes concurrent requests which #8 tried to implement, even took some ideas from the feedback of that pr.

jhodapp commented:
Hi @AntoniosBarotsis, I've been limited on time so I will do a little bit of this review at a time. Thank you for this PR!

Some first overall thoughts. Would you mind adding a more specific PR title, directly stating which flags you're adding to the CLI?

Were the dependency updates strictly required for your new functionality that you added?

AntoniosBarotsis commented:
Hey! no worries!

I mention the flags in the linked issue [here](https://github.com/Jim-Hodapp-Coaching/ambi_mock_client/issues/13) but I guess I could add them in the PR description as well.

As for the updates, I generally like having up to date dependencies in general as long as they are not breaking anything (which is the case here). Some of them had documented vulnerabilities as well which isn't a big problem here but figured why not. The main upgrade that mattered was `clap` because version 3 to 4 had a bunch of breaking changes which could hurt the development of the project in the future. Everything else was minor updates only.

jhodapp commented:
@AntoniosBarotsis Thanks for this, it's ok for this PR, especially since we didn't mention this ahead of time, but generally it's a helpful practice to keep your PR focused on a specific objective. Updating the dependencies would be considered a separate focus that would help make your PR shorter and easier to digest for reviewers. But like I said, it's fine for this particular PR, just keep this in mind for future ones.

AntoniosBarotsis commented:
I think in this case it's a bit more trivial as all of the stuff I upgraded are minor versions of popular crates which you can confidently assume have no breaking changes or anything but more importantly because this project is a standalone binary. If it was a library (especially one in an embedded context like the wifi drivers you are developing for instance) I would've definitely not made these so lightly but in this case, I do believe they are somewhat trivial.

That said, I should've definitely mentioned my reasoning as to why I made them in the desc PR's description, thanks for pointing that out!

jhodapp commented:
Yeah you're welcome, agreed that it's pretty trivial in this case which is why I am not asking that you remove them from this PR and out into a separate one. Thanks for understanding.

More review on the content of your PR soon...

calebbourg commented on src/data.rs:
```diff
@@ -0,0 +1,107 @@
+impl AirPurity {
+    pub(crate) fn from_value(value: f64) -> AirPurity {
+        match value {
+            value if (f64::MIN..=50.0).contains(&value) => AirPurity::FreshAir,
```
I wonder if just using the same pattern you've used lower 

```suggestion
            value if value < 50.0 => AirPurity::FreshAir,
```
Would be more performant ? Not sure if this would need to check the whole range here or not. In any case keeping consistency in the pattern might be nice. Not a huge deal by any means.

calebbourg commented on src/data.rs:
```diff
@@ -0,0 +1,107 @@
+            AirPurity::Low => write!(f, "Fresh Air"),
+            AirPurity::High => write!(f, "Low Pollution"),
+            AirPurity::Dangerous => write!(f, "High Pollution"),
+            AirPurity::FreshAir => write!(f, "Dangerous Pollution"),
```
Should this be `"Fresh Air"` ?

AntoniosBarotsis replied:
Oh nice catch. This is not mine, I just moved it around, you can see it in master [here](https://github.com/Jim-Hodapp-Coaching/ambi_mock_client/blob/9937b55c33869852513441a5c2aee307b2f81ed0/src/lib.rs#L84-L93). What should I name it instead?

calebbourg replied:
I'm guessing I added that :)

Actually looking at it now it's all a bit off... probably also my fault

What do you think about?
```rust
AirPurity::Low => write!(f, "Low Pollution"),
AirPurity::High => write!(f, "High Pollution"),
AirPurity::Dangerous => write!(f, "Dangerous Pollution"),
AirPurity::FreshAir => write!(f, "Fresh Air"),
```

AntoniosBarotsis replied:
Makes more sense, will add that!

calebbourg commented on src/data.rs:
```diff
@@ -0,0 +1,107 @@
+    }
+}
+
+pub(crate) fn random_gen_humidity() -> f64 {
```
You could pass `rng` in as an argument for these functions so that you don't have to generate one for each function call. Then update the caller of these functions to invoke `thread_rng()` and pass in the value.

calebbourg commented on src/requests.rs:
```diff
@@ -0,0 +1,236 @@
+            (Some(time_per_request), Some(_)) => *time_per_request,
+        };
+
+        let num_threads = match self.num_threads {
+            Some(num_threads) => num_threads,
+            None => DEFAULT_NUM_THREADS,
+        };
```
```suggestion
        let num_threads = self.num_threads.unwrap_or(DEFAULT_NUM_THREADS);
```

calebbourg commented on src/requests.rs:
```diff
@@ -0,0 +1,236 @@
+        };
+
+        // Make sure that the number of threads is in [1, `MAX_NUM_THREADS`].
+        match num_threads {
```
Is it possible to handle this sort of validation with Clap directly?

AntoniosBarotsis replied:
[It turns out it is yes](https://docs.rs/clap/latest/clap/error/enum.ErrorKind.html#variant.ValueValidation), didn't know :)

calebbourg reviewed (commented):
@AntoniosBarotsis This looks great! Thank you for adding this. Just left a few questions/ideas.

AntoniosBarotsis commented:
@calebbourg Just pushed the changes, let me know if everything is resolved properly :)

calebbourg commented:
Love it!

jhodapp commented on Cargo.toml:
```diff
@@ -1,19 +1,24 @@
 [package]
 name = "ambi_mock_client"
+authors = ["Rust Never Sleeps community (https://github.com/Jim-Hodapp-Coaching/)"]
```
Thanks for adding this!

jhodapp commented on src/lib.rs:
```diff
@@ -9,181 +9,86 @@
-    #[clap(short, long)]
+    #[arg(short, long)]
     pub debug: bool,
+    /// The amount of requests to make
```
```suggestion
    /// The number of sensor readings to post.
```

jhodapp commented on src/lib.rs:
```diff
@@ -9,181 +9,86 @@
+    /// [DEFAULT: 1]
+    #[arg(short = 'n', long)]
+    pub request_amount: Option<u32>,
+    /// The time between each request (in seconds)
```
```suggestion
    /// The time between each sensor reading post (in seconds).
```

jhodapp commented on src/lib.rs:
```diff
@@ -9,181 +9,86 @@
+    /// [DEFAULT: 10]
+    #[arg(short = 't', long = "time-per-request")]
+    pub time_per_request_s: Option<u64>,
+    /// The total time over which all the requests must be sent (in seconds, alternative to -t)
```
```suggestion
    /// The total time over which all the sensor reading posts must be sent (in seconds, alternative to -t).
```

jhodapp commented on src/lib.rs:
```diff
@@ -9,181 +9,86 @@
+    /// The total time over which all the requests must be sent (in seconds, alternative to -t)
+    #[arg(short = 'T', long = "total-time")]
+    pub total_time_s: Option<u64>,
+    /// The number of threads to spawn. The workload will be cloned to each thread, not divided
```
```suggestion
    /// The number of threads to spawn. The workload will be cloned to each thread, not divided.
```

jhodapp commented on src/requests.rs:
```diff
@@ -0,0 +1,229 @@
+    num_threads: Option<u32>,
+}
+
+impl RequestSchedulerBuilder {
```
May I suggest that we rename from using the word "Request" to "SensorPost" or something like it. For me, request is a little misleading on the intended functionality here, as it's about posting data to a backend and not requesting data from it. Thoughts?

AntoniosBarotsis replied:
So should I rename it to `SensorPostSchedulerBuilder`? What about all the other mentions of the word `request` in the file?

I feel like "request" does indeed not make too much sense here.

jhodapp replied:
Maybe something like `PostSchedulerBuilder` would be adequate unless you think we might have different types of schedulers?

jhodapp commented on src/lib.rs:
```diff
@@ -9,181 +9,86 @@
 
-pub fn run(cli: &Cli) {
-    println!("\r\ncli: {:?}\r\n", cli);
+    let req_scheduler = RequestSchedulerBuilder::default()
```
I love this, great use of the builder pattern!

jhodapp reviewed (changes requested):
Looking great, made a few recommendations inline. Open to discussing any of them.

jhodapp commented on src/requests.rs:
```diff
@@ -0,0 +1,229 @@
+
+#[derive(Clone, Copy)]
+pub struct PostSchedulerBuilder {
+    request_amount: Option<u32>,
```
@AntoniosBarotsis I think we should also change nearly ever mention of a `request` to be `post` or something contextually relevant. Does it even make sense to have a `requests.rs` now? Consider changing it to `sensor_posts.rs` as on idea.

AntoniosBarotsis replied:
I've left a few mentions of `request` where I feel like it makes sense, such as [here](https://github.com/Jim-Hodapp-Coaching/ambi_mock_client/blob/5782646641471f87fc0b032e2d9c2b8ddb1e89aa/src/sensor_posts.rs#L181-L186). Let me know what you think.

AntoniosBarotsis replied:
@jhodapp, probably missed my comment here :)

jhodapp reviewed (changes requested):
Definitely a step in the right direction, made a suggestion to further rename things.

AntoniosBarotsis commented:
I added this section [here](https://github.com/Jim-Hodapp-Coaching/ambi_mock_client/blob/863425fb6a47f69fbce63f5cd34ab91bc33c8b94/src/sensor_posts.rs#L165-L173) which will make sure to log the response message as an error if the response code is not in the 200-299 range, makes it easier to immediately identify if something did not work.

jhodapp reviewed (changes requested):
Looks like the tests don't currently pass, I'm seeing 2 failures. Are you seeing these as well?

```sh
running 4 tests
test sensor_posts::tests::test_invalid_num_threads_low ... FAILED
test sensor_posts::tests::test_invalid_num_threads_high ... FAILED
test data::tests::air_purity_from_value_returns_correct_enum ... ok
test tests::verify_cli ... ok

failures:

---- sensor_posts::tests::test_invalid_num_threads_low stdout ----
thread 'sensor_posts::tests::test_invalid_num_threads_low' panicked at 'assertion failed: req_scheduler.is_err()', src/sensor_posts.rs:224:9

---- sensor_posts::tests::test_invalid_num_threads_high stdout ----
thread 'sensor_posts::tests::test_invalid_num_threads_high' panicked at 'assertion failed: req_scheduler.is_err()', src/sensor_posts.rs:233:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


failures:
    sensor_posts::tests::test_invalid_num_threads_high
    sensor_posts::tests::test_invalid_num_threads_low

test result: FAILED. 2 passed; 2 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
```

AntoniosBarotsis commented:
Ah, should've run those again (probably a good enough excuse to set up a basic CI for this). 

Caleb suggested I move some of the validation (for what is considered a valid number of threads) out of the code and use Clap's validators instead. The tests *don't* fail anymore because the check would normally happen higher up the stack (as in, if you indeed try to run it with 0 or 11 threads which is what the tests are doing, it will fail with an error as it should). 

I'll fix the tests later today and make sure they run in GitHub actions from now on.

jhodapp commented:
Sounds great, I look forward to seeing your updates!

AntoniosBarotsis commented:
Ok so, fixed the tests, they look like [this](https://github.com/Jim-Hodapp-Coaching/ambi_mock_client/blob/f8510ad54997403e65ed223c1d93e16c18fed517/src/lib.rs#L95-L112) now:

```rs
let cli = Cli::try_parse_from(vec!["", "-p", "0"]);
assert!(cli.is_err());

let cli_err = cli.unwrap_err();
assert_eq!(cli_err.kind(), ErrorKind::ValueValidation);
```

I just ensure that clap is *failing* to parse the input as we consider this amount of threads invalid.

I also added the basic workflow I mentioned earlier although I believe it won't run until it's merged which is a bit annoying. 

Sorry for the many back and forths, last request review is [this](https://github.com/Jim-Hodapp-Coaching/ambi_mock_client/pull/15#discussion_r1127075585) (waiting for reply) and everything should be fine after that!

jhodapp reviewed (approved):
Looks great Tony, thanks again for this! Tests pass for me now.
